Sources/Core/Client/WebSocket/WebSocket.swift

```diff
@@ -161,11 +177,87 @@ public final class WebSocket {
         }
     }
     
-    func clearStateAfterDisconnect() {
+    private func isTokenExpired() -> Bool {
+        guard let lastJSONError = lastJSONError else {
+            return false
+        }
+        
+        let isTokenExpired = lastJSONError.code == ClientErrorResponse.tokenExpiredErrorCode
+        return isTokenExpired && Client.shared.touchTokenProvider()
```

Comment:
We really need dependency injection.. `WebSocket` instance method mutating singleton `Client` is unexpected

Reply:
How do you suggest to inject it?

Reply:
Initializer?

Reply:
`Client` is a shared instance.

Reply:
Why do you want to inject WebSocket?

Reply:
I want to inject `Client` into `WebSocket` (or anywhere where it's used) so that testing is easier and it's easier to follow & understand. Currently, `Client.shared` is accessed from multiple places which makes it hard to follow where it's mutated from.

Reply:
It's probably not for this PR, though, out of scope

Reply:
`Client` use `WebSocket`. `WebSocket` _doesn't know_ about the `Client`.

Reply:
It does. Here, in an instance method of `WebSocket`, `Client.shared` is accessed

Reply:
It's only used for logger, that easy to avoid. I mean in general functionality.